### PR TITLE
chore: prove that subtraction on BitStreams corresponds to subtraction on BitVectors

### PR DESCRIPTION
**Pull request by AtticusKuhn**
This PR proves the key theorem

```lean
theorem ofBitVec_sub : ofBitVec (x - y) ≈ʷ (ofBitVec x) - (ofBitVec y)  := by
  calc
  _ ≈ʷ ofBitVec (x + - y) := by rw [BitVec.sub_add_neg]
  _ ≈ʷ ofBitVec x + ofBitVec (-y) := ofBitVec_add
  _ ≈ʷ ofBitVec x + - ofBitVec y := add_congr equal_up_to_refl ofBitVec_neg
  _ ≈ʷ ofBitVec x - ofBitVec y := by rw [sub_add_neg]
```

Which shows that subtraction on BitStreams corresponds to subtraction on BitVectors.

In doing so, this PR removes the last "sorry" from the file BitStream.lean.

Note that this PR is a continuation of PR
https://github.com/opencompl/lean-mlir/pull/554

### CODE REVIEW

File: SSA/Experimental/Bits/Fast/BitStream.lean [hunk 1]
```diff
@@ -474,6 +522,13 @@ theorem ofBitVec_neg : ofBitVec (- x) ≈ʷ - (ofBitVec x) := by
   _ ≈ʷ ~~~ ofBitVec x   + 1            := add_congr ofBitVec_not_eqTo ofBitVec_one_eqTo_ofNat
   _ ≈ʷ - (ofBitVec x)                  := by rw [neg_eq_not_add]
 
+theorem ofBitVec_sub : ofBitVec (x - y) ≈ʷ (ofBitVec x) - (ofBitVec y)  := by
+  calc
+  _ ≈ʷ ofBitVec (x + - y) := by rw [BitVec.sub_add_neg]
+  _ ≈ʷ ofBitVec x + ofBitVec (-y) := ofBitVec_add
+  _ ≈ʷ ofBitVec x + - ofBitVec y := add_congr equal_up_to_refl ofBitVec_neg
+  _ ≈ʷ ofBitVec x - ofBitVec y := by rw [sub_add_neg]
```
Comment (alexkeizer) [inline SSA/Experimental/Bits/Fast/BitStream.lean]:
```suggestion
  _ ≈ʷ ofBitVec (x + -y) := by rw [BitVec.sub_add_neg]
  _ ≈ʷ ofBitVec x + ofBitVec (-y) := ofBitVec_add
  _ ≈ʷ ofBitVec x + -(ofBitVec y) := add_congr equal_up_to_refl ofBitVec_neg
  _ ≈ʷ ofBitVec x - ofBitVec y := by rw [sub_add_neg]
```
Minor spacing nit

File: SSA/Experimental/Bits/Fast/BitStream.lean [hunk 2]
```diff
@@ -367,28 +367,60 @@ variable {w : Nat} {x y : BitVec w} {a b a' b' : BitStream}
...
+theorem carry_succ_left {a b : BitStream} (i : ℕ) :
+    xor (b (i + 1)) (a.subCarries? b i) = ((!b (i + 1)) != ((b.negAux i).2 != (a.addAux (fun n => (b.negAux n).1) i).2)) := by
+  simp [Bool.xor_shift, subCarries?_correct i]
+
+theorem sub_add_lemma (i : Nat) :
+    a.subAux b i = ⟨(a.addAux b.neg i).1, subCarries? a b i⟩  := by
+  induction' i with i ih
+  · simp [subAux,addAux,negAux, BitVec.adcb, subCarries?, neg]
+  · simp [subAux,addAux,negAux, BitVec.adcb, ih, carry_succ_left i, subCarries?, neg]
+    rfl
```
Comment (tobiasgrosser) [inline SSA/Experimental/Bits/Fast/BitStream.lean]:
```suggestion
  induction' i with i ih
  · simp [subAux, addAux, negAux, BitVec.adcb, subCarries?, neg]
  · simp [subAux, addAux, negAux, BitVec.adcb, ih, carry_succ_left i, subCarries?, neg]
    rfl
```

File: SSA/Experimental/Bits/Fast/BitStream.lean [hunk 3]
```diff
@@ -437,7 +473,6 @@ theorem add_congr (e1 : a ≈ʷ b) (e2 : c ≈ʷ d) : (a + c) ≈ʷ (b + d) := b
     · simp only [addAux, e1 _ h, e2 _ h, ih (by omega)]
   simp [HAdd.hAdd, Add.add, BitStream.add, add_congr_lemma, addAux]
 
-
 theorem not_congr (e1 : a ≈ʷ b) : (~~~a) ≈ʷ ~~~b := by
```
Comment (tobiasgrosser) [inline SSA/Experimental/Bits/Fast/BitStream.lean]: This is unrelated. Drop this change?

File: SSA/Projects/InstCombine/ForLean.lean [hunk 4]
```diff
@@ -590,3 +590,16 @@ theorem Bool.xor_decide (p q : Prop) [dp : Decidable p] [Decidable q] :
...
+  <;> cases b
+  <;> simp
+
+theorem Bool.xor_xor_eq_not {a b : Bool} : xor (!xor a b) b = !a := by
+  cases a
+  <;> simp
+
+theorem Bool.xor_and_eq_and {a b : Bool} : (!xor a b && b) = (a && b) := by
+  cases a
+  <;> simp
```
Comment (tobiasgrosser) [inline SSA/Projects/InstCombine/ForLean.lean]:
```suggestion
theorem Bool.xor_not_xor {a b : Bool} : xor (!xor a b) b = !a := by
  cases a
  <;> simp

theorem Bool.not_xor_and_self {a b : Bool} : (!xor a b && b) = (a && b) := by
  cases a
  <;> simp
```

I am not too sure about these names, but maybe check what lean is doing in its core library. Or maybe @alexkeizer has naming advise.

File: SSA/Projects/InstCombine/ForLean.lean [hunk 5]
```diff
@@ -590,3 +590,16 @@ theorem Bool.xor_decide (p q : Prop) [dp : Decidable p] [Decidable q] :
     (decide p).xor (decide q) = decide (p ≠ q) := by
   cases' dp with pt pt
   <;> simp [pt]
+
+theorem Bool.xor_shift {a b c : Bool} : xor a b = c ↔ b = xor a c := by
+  cases a
```
Comment (tobiasgrosser) [inline SSA/Projects/InstCombine/ForLean.lean]: Why is this named shift? Is there precedence somewhere in the lean codebase and/or mathlib?

File: SSA/Experimental/Bits/Fast/BitStream.lean [hunk 6]
```diff
@@ -367,28 +367,60 @@ variable {w : Nat} {x y : BitVec w} {a b a' b' : BitStream}
...
+  simp [Bool.xor_shift, subCarries?_correct i]
+
+theorem sub_add_lemma (i : Nat) :
+    a.subAux b i = ⟨(a.addAux b.neg i).1, subCarries? a b i⟩  := by
+  induction' i with i ih
+  · simp [subAux,addAux,negAux, BitVec.adcb, subCarries?, neg]
+  · simp [subAux,addAux,negAux, BitVec.adcb, ih, carry_succ_left i, subCarries?, neg]
+    rfl
+
+theorem sub_add_neg {a b : BitStream} : a - b = a + (-b) := by
```
Comment (tobiasgrosser) [inline SSA/Experimental/Bits/Fast/BitStream.lean]:
```suggestion
theorem sub_eq_add_neg {a b : BitStream} : a - b = a + (-b) := by
```

Comment (tobiasgrosser) [reply SSA/Experimental/Bits/Fast/BitStream.lean]: Maybe check in lean if there is a naming example, e.g., for Nat or sth similar.

File: SSA/Experimental/Bits/Fast/BitStream.lean [hunk 7]
```diff
@@ -367,28 +367,60 @@ variable {w : Nat} {x y : BitVec w} {a b a' b' : BitStream}
...
     induction' i with i ih
     · simp [neg, negAux]
-    · simp [neg, negAux, ih, xor_xor_eq_not, xor_and_eq_and]
+    · simp [neg, negAux, ih, Bool.xor_xor_eq_not, Bool.xor_and_eq_and]
   simp [Neg.neg, neg, neg_lemma]
 
--- TODO: This sorry is difficult, and will be proven in a later Pull Request.
-theorem ofBitVec_sub : ofBitVec (x - y) ≈ʷ (ofBitVec x) - (ofBitVec y)  := by
-  sorry
+def subCarries? (a b : BitStream) (i : Nat) : Bool :=
```
Comment (tobiasgrosser) [inline SSA/Experimental/Bits/Fast/BitStream.lean:329]: I feel this and the next four lemmas could benefit from a comment. Otherwise, these lemmas are really hard to understand.

File: SSA/Experimental/Bits/Fast/BitStream.lean (hunk 6 above)
Comment (AtticusKuhn) [reply SSA/Experimental/Bits/Fast/BitStream.lean]: Yes, there is 
```lean
theorem Int.sub_eq_add_neg {a : Int} {b : Int} : a - b = a + -b	
```

**Comment by tobiasgrosser** [comment]
@AtticusKuhn, what a success! 🎉  Turning my broken hack into such a beautiful proof.

I left many comments, but this is one of the most polished PRs I started reviewing. In retrospect, the proof seems straightforward, which is remarkable given the complex proof states one can reach. You broke it down into a couple of well-structured semantic pieces.

**Review by tobiasgrosser — APPROVED** [review]

File: SSA/Projects/InstCombine/ForLean.lean (hunk 5 above)
Comment (bollu) [reply SSA/Projects/InstCombine/ForLean.lean]: I'd call it `xor_inv_left`, because it's saying that $$a \oplus b = c$$ iff $$a \oplus a \oplus b = a \oplus c$$, which implies $$b = a \oplus c$$ (modulo 2,  $$a = a^{-1}$$)

File: SSA/Experimental/Bits/Fast/BitStream.lean [hunk 8]
```diff
@@ -373,52 +373,55 @@ theorem neg_neg : a = - - a := by
...
+  · simp [negAux, addAux, BitVec.adcb, neg]
+    cases' ih with l l
+    <;> cases b (i + 1)
+    <;> cases a (i + 1)
+    <;> simp [l]
 
+/--
+  Whether a - b will overflow is equivalent to
+  -b overflows = (a + - b) overflows
+-/
```
Comment (tobiasgrosser) [inline SSA/Experimental/Bits/Fast/BitStream.lean]:
```suggestion
  Whether a - b will overflow is equivalent to -b overflows = (a + - b) overflows.
-/
```

File: SSA/Experimental/Bits/Fast/BitStream.lean [hunk 9]
```diff
@@ -367,28 +367,63 @@ variable {w : Nat} {x y : BitVec w} {a b a' b' : BitStream}
...
+theorem sub_add_lemma (i : Nat) :
+    a.subAux b i = ⟨(a.addAux b.neg i).1, subCarries? a b i⟩ := by
+  induction' i with i ih
+  · simp [subAux, addAux, negAux, BitVec.adcb, subCarries?, neg]
+  · simp [subAux, addAux, negAux, BitVec.adcb, ih, Bool.xor_neq_self, subCarries?, neg, Bool.xor_inv_left, subCarries?_correct i]
+
+
+theorem sub_eq_add_neg : a - b = a + (-b) := by
+  ext i
+  simp [HAdd.hAdd, HSub.hSub, Neg.neg, Sub.sub, BitStream.sub, Add.add, BitStream.add, sub_add_lemma i]
```
Comment (tobiasgrosser) [inline SSA/Experimental/Bits/Fast/BitStream.lean]:
```suggestion
  · simp [subAux, addAux, negAux, BitVec.adcb, ih, Bool.xor_neq_self, subCarries?, neg, Bool.xor_inv_left, subCarries?_correct i]

theorem sub_eq_add_neg : a - b = a + (-b) := by
  ext i
  simp [HAdd.hAdd, HSub.hSub, Neg.neg, Sub.sub, BitStream.sub, Add.add, BitStream.add, sub_add_lemma i]
```

File: SSA/Experimental/Bits/Fast/BitStream.lean (hunk 3 above)
Comment (tobiasgrosser) [reply SSA/Experimental/Bits/Fast/BitStream.lean]: This dropped line is still there. I suggest to do this cleanup in a separate PR (no review required).

File: SSA/Experimental/Bits/Fast/BitStream.lean [hunk 10]
```diff
@@ -367,28 +367,63 @@ variable {w : Nat} {x y : BitVec w} {a b a' b' : BitStream}
...
-theorem xor_xor_eq_not {a b : Bool} : xor (!xor a b) b = !a := by
-  cases a
-  <;> cases b
-  <;> simp
-
-theorem xor_and_eq_and {a b : Bool} : (!xor a b && b) = (a && b) := by
-  cases a
-  <;> cases b
-  <;> simp
-
```
Comment (tobiasgrosser) [inline SSA/Experimental/Bits/Fast/BitStream.lean]: I would suggest you move these statements into ForLean in a separate PR (no review required).

File: SSA/Experimental/Bits/Fast/BitStream.lean [hunk 11]
```diff
@@ -367,28 +367,60 @@ variable {w : Nat} {x y : BitVec w} {a b a' b' : BitStream}
...
+theorem subCarries?_correct (i : Nat) :
+    a.subCarries? b i = ((b.negAux i).2 == (a.addAux b.neg i).2) := by
+  induction' i with i ih
+  · simp [subCarries?, negAux, addAux, BitVec.adcb, neg]
+  · by_cases a1 : a (i + 1)
+    <;> by_cases b1 : b (i + 1)
+    <;> cases' @neg_or_add a b i with h h
+    <;> simp [h, subCarries?, ih, negAux, addAux, BitVec.adcb, neg, a1, b1]
+
+theorem sub_add_lemma (i : Nat) :
```
Comment (tobiasgrosser) [inline SSA/Experimental/Bits/Fast/BitStream.lean]: This name could also use some thought?

**Comment by AtticusKuhn** [comment]
CC: @bollu @alexkeizer 

I think this PR is in pretty good shape. I would appreciate if you would review it.

File: SSA/Experimental/Bits/Fast/BitStream.lean (hunk 10 above)
Comment (tobiasgrosser) [reply SSA/Experimental/Bits/Fast/BitStream.lean]: ... such that you can then rebase this PR on top of the cleanup, which reduces the size of this PR.

File: SSA/Experimental/Bits/Fast/BitStream.lean [hunk 12]
```diff
@@ -429,6 +461,10 @@ instance congr_equiv : Equivalence (EqualUpTo w) where
   symm := equal_up_to_symm
   trans := equal_up_to_trans
 
+theorem BitVec.sub_eq_add_neg : x - y = x + (- y) := by
+  simp only [HAdd.hAdd, HSub.hSub, Neg.neg, Sub.sub, BitVec.sub, Add.add, BitVec.add]
+  simp [BitVec.ofNat, Fin.ofNat', add_comm]
```
Comment (tobiasgrosser) [inline SSA/Experimental/Bits/Fast/BitStream.lean]: Should this move into `ForLean.lean` and have there an explicit `{x y : BitVec w}` type annotation.

File: SSA/Projects/InstCombine/ForLean.lean [hunk 13]
```diff
@@ -590,3 +590,30 @@ theorem Bool.xor_decide (p q : Prop) [dp : Decidable p] [Decidable q] :
     (decide p).xor (decide q) = decide (p ≠ q) := by
   cases' dp with pt pt
   <;> simp [pt]
+
```
Comment (tobiasgrosser) [inline SSA/Projects/InstCombine/ForLean.lean]: When you move the unrelated Boolean statements to `ForLean.lean`, you could also drop the `Bool.` prefix and create a section instead.

File: SSA/Projects/InstCombine/ForLean.lean [hunk 14]
```diff
@@ -590,3 +590,30 @@ theorem Bool.xor_decide (p q : Prop) [dp : Decidable p] [Decidable q] :
...
+theorem Bool.xor_inv_left {a b c : Bool} : xor a b = c ↔ b = xor a c := by
+  cases a
+  <;> cases b
+  <;> simp
+
+theorem Bool.xor_not_xor {a b : Bool} : xor (!xor a b) b = !a := by
+  cases a
+  <;> simp
+
+theorem Bool.not_xor_and_self {a b : Bool} : (!xor a b && b) = (a && b) := by
```
Comment (tobiasgrosser) [inline SSA/Projects/InstCombine/ForLean.lean]: Should this be `@[simp]`?

File: SSA/Projects/InstCombine/ForLean.lean [hunk 15]
```diff
@@ -590,3 +590,30 @@ theorem Bool.xor_decide (p q : Prop) [dp : Decidable p] [Decidable q] :
     (decide p).xor (decide q) = decide (p ≠ q) := by
   cases' dp with pt pt
   <;> simp [pt]
+
+theorem Bool.xor_inv_left {a b c : Bool} : xor a b = c ↔ b = xor a c := by
+  cases a
+  <;> cases b
+  <;> simp
+
+theorem Bool.xor_not_xor {a b : Bool} : xor (!xor a b) b = !a := by
```
Comment (tobiasgrosser) [inline SSA/Projects/InstCombine/ForLean.lean]: Should this be `@[simp]`?

File: SSA/Projects/InstCombine/ForLean.lean [hunk 16]
```diff
@@ -590,3 +590,30 @@ theorem Bool.xor_decide (p q : Prop) [dp : Decidable p] [Decidable q] :
...
+  cases a
+  <;> simp
+
+theorem Bool.xor_neq_self {a c : Bool} :  xor a ((!a) != c) = !c := by
+  cases a
+  <;> simp
+
+@[simp]
+theorem Bool.not_eq_and {a b : Bool}: ((!b) == (a && b)) = (!a && b)  := by
+  cases a
```
Comment (tobiasgrosser) [inline SSA/Projects/InstCombine/ForLean.lean]:
```suggestion
theorem Bool.not_eq_and {a b : Bool} : ((!b) == (a && b)) = (!a && b)  := by
  cases a
```

Are all the parenthesis needed here?

File: SSA/Projects/InstCombine/ForLean.lean [hunk 17]
```diff
@@ -590,3 +590,30 @@ theorem Bool.xor_decide (p q : Prop) [dp : Decidable p] [Decidable q] :
...
+  <;> simp
+
+@[simp]
+theorem Bool.not_neq {a b : Bool}:  (!(bne a b)) = (a == b) := by
+  cases a
```
Comment (tobiasgrosser) [inline SSA/Projects/InstCombine/ForLean.lean]:
```suggestion
theorem Bool.not_neq {a b : Bool} : (!(bne a b)) = (a == b) := by
  cases a
```
Are all the parenthesis needed here?

File: SSA/Projects/InstCombine/ForLean.lean [hunk 18]
```diff
@@ -590,3 +590,30 @@ theorem Bool.xor_decide (p q : Prop) [dp : Decidable p] [Decidable q] :
...
+
+theorem Bool.xor_not_xor {a b : Bool} : xor (!xor a b) b = !a := by
+  cases a
+  <;> simp
+
+theorem Bool.not_xor_and_self {a b : Bool} : (!xor a b && b) = (a && b) := by
+  cases a
+  <;> simp
+
+theorem Bool.xor_neq_self {a c : Bool} :  xor a ((!a) != c) = !c := by
```
Comment (tobiasgrosser) [inline SSA/Projects/InstCombine/ForLean.lean]: Are all the parenthesis needed here?

File: SSA/Projects/InstCombine/ForLean.lean [hunk 19]
```diff
@@ -596,24 +596,31 @@ theorem Bool.xor_inv_left {a b c : Bool} : xor a b = c ↔ b = xor a c := by
...
 
 @[simp]
-theorem Bool.not_neq {a b : Bool}:  (!(bne a b)) = (a == b) := by
+theorem Bool.not_neq {a b : Bool} :  (!bne a b) = (a == b) := by
   cases a
   <;> simp
+
+theorem BitVec.sub_eq_add_neg {w : Nat} {x y : BitVec w} : x - y = x + (- y) := by
+  simp only [HAdd.hAdd, HSub.hSub, Neg.neg, Sub.sub, BitVec.sub, Add.add, BitVec.add]
+  simp [BitVec.ofNat, Fin.ofNat', add_comm]
```
Comment (tobiasgrosser) [inline SSA/Projects/InstCombine/ForLean.lean]: This belongs higher up into the `BitVec` section of this file.

File: SSA/Projects/InstCombine/ForLean.lean [hunk 20]
```diff
@@ -590,3 +590,37 @@ theorem Bool.xor_decide (p q : Prop) [dp : Decidable p] [Decidable q] :
...
+  <;> simp
+
+@[simp]
+theorem Bool.not_eq_and {a b : Bool} : ((!b) == (a && b)) = (!a && b)  := by
+  cases a
+  <;> simp
+
+@[simp]
+theorem Bool.not_neq {a b : Bool} :  (!bne a b) = (a == b) := by
+  cases a
```
Comment (tobiasgrosser) [inline SSA/Projects/InstCombine/ForLean.lean]:
```suggestion
theorem Bool.not_neq {a b : Bool} : (!bne a b) = (a == b) := by
  cases a
```

File: SSA/Projects/InstCombine/ForLean.lean [hunk 21]
```diff
@@ -590,3 +590,37 @@ theorem Bool.xor_decide (p q : Prop) [dp : Decidable p] [Decidable q] :
...
+  <;> simp
+
+@[simp]
+theorem Bool.not_xor_and_self {a b : Bool} : (!xor a b && b) = (a && b) := by
+  cases a
+  <;> simp
+
+@[simp]
+theorem Bool.xor_neq_self {a c : Bool} :  xor a ((!a) != c) = !c := by
```
Comment (tobiasgrosser) [inline SSA/Projects/InstCombine/ForLean.lean]:
```suggestion
@[simp]
theorem Bool.xor_neq_self {a c : Bool} : xor a ((!a) != c) = !c := by
```

Comment (tobiasgrosser) [reply SSA/Projects/InstCombine/ForLean.lean]: Also, you probably want to use `a` and `b` as variable names.

File: SSA/Projects/InstCombine/ForLean.lean [hunk 22]
```diff
@@ -178,6 +178,10 @@ lemma mul_eq_and (a b : BitVec 1) : a * b = a &&& b := by
   have hb : b = 0 ∨ b = 1 := width_one_cases _
   rcases ha with h | h <;> (rcases hb with h' | h' <;> (simp [h, h']))
 
+theorem sub_eq_add_neg {w : Nat} {x y : BitVec w} : x - y = x + (- y) := by
```
Comment (tobiasgrosser) [inline SSA/Projects/InstCombine/ForLean.lean:181]:
```suggestion
theorem add_neg {w : Nat} {x y : BitVec w} : x + (-y) = x - y := by
```

Check for similar naming in `Int.` or `Nat.` to confirm that this is the canonical form.

Comment (tobiasgrosser) [reply SSA/Projects/InstCombine/ForLean.lean:181]: We probably also want to make this a `@[simp]`.

Comment (AtticusKuhn) [reply SSA/Projects/InstCombine/ForLean.lean:181]: https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Group/Defs.html#sub_eq_add_neg

Comment (AtticusKuhn) [reply SSA/Projects/InstCombine/ForLean.lean:181]: https://github.com/leanprover/lean4/blob/0edf1bac392f7e2fe0266b28b51c498306363a84/src/Init/Data/Int/Lemmas.lean#L87-L87

File: SSA/Projects/InstCombine/ForLean.lean [hunk 23]
```diff
@@ -590,3 +594,33 @@ theorem Bool.xor_decide (p q : Prop) [dp : Decidable p] [Decidable q] :
...
+  <;> simp
+
+@[simp]
+theorem Bool.not_xor_and_self {a b : Bool} : (!xor a b && b) = (a && b) := by
+  cases a
+  <;> simp
+
+@[simp]
+theorem Bool.xor_neq_self {a b : Bool} : xor a ((!a) != b) = !b := by
+  cases a
```
Comment (tobiasgrosser) [inline SSA/Projects/InstCombine/ForLean.lean]:
```suggestion
theorem Bool.xor_ne_self {a b : Bool} : xor a ((!a) != b) = !b := by
  cases a
```

File: SSA/Projects/InstCombine/ForLean.lean [hunk 24]
```diff
@@ -590,3 +594,33 @@ theorem Bool.xor_decide (p q : Prop) [dp : Decidable p] [Decidable q] :
...
+  <;> simp
+
+@[simp]
+theorem Bool.not_eq_and {a b : Bool} : ((!b) == (a && b)) = (!a && b)  := by
+  cases a
+  <;> simp
+
+@[simp]
+theorem Bool.not_neq {a b : Bool} : (!bne a b) = (a == b) := by
+  cases a
```
Comment (tobiasgrosser) [inline SSA/Projects/InstCombine/ForLean.lean]:
```suggestion
theorem Bool.not_ne {a b : Bool} : (!bne a b) = (a == b) := by
  cases a
```

File: SSA/Projects/InstCombine/ForLean.lean [hunk 25]
```diff
@@ -605,4 +609,24 @@ theorem not_xor_and_self {a b : Bool} : (!xor a b && b) = (a && b) := by
...
   <;> simp
 
+theorem xor_inv_left {a b c : Bool} : xor a b = c ↔ b = xor a c := by
+  cases a
+  <;> cases b
+  <;> simp
+
+@[simp]
+theorem xor_neq_self {a b : Bool} : xor a ((!a) != b) = !b := by
+  cases a
```
Comment (tobiasgrosser) [inline SSA/Projects/InstCombine/ForLean.lean]:
```suggestion
theorem xor_ne_self {a b : Bool} : xor a ((!a) != b) = !b := by
  cases a
```

File: SSA/Projects/InstCombine/ForLean.lean [hunk 26]
```diff
@@ -605,4 +609,24 @@ theorem not_xor_and_self {a b : Bool} : (!xor a b && b) = (a && b) := by
...
+  <;> simp
+
+@[simp]
+theorem not_eq_and {a b : Bool} : ((!b) == (a && b)) = (!a && b)  := by
+  cases a
+  <;> simp
+
+@[simp]
+theorem not_neq {a b : Bool} : (!bne a b) = (a == b) := by
```
Comment (tobiasgrosser) [inline SSA/Projects/InstCombine/ForLean.lean]:
```suggestion
theorem not_ne {a b : Bool} : (!bne a b) = (a == b) := by
```

**Review by tobiasgrosser — APPROVED** [review]
I am done. This looks perfect to me. @alexkeizer @bollu, feedback would be appreciated.

File: SSA/Projects/InstCombine/ForLean.lean (hunk 16 above)
Comment (tobiasgrosser) [reply SSA/Projects/InstCombine/ForLean.lean]: I think this has meanwhile also been resolved.

File: SSA/Projects/InstCombine/ForLean.lean (hunk 13 above)
Comment (tobiasgrosser) [reply SSA/Projects/InstCombine/ForLean.lean]: I implemented this.

File: SSA/Experimental/Bits/Fast/BitStream.lean (hunk 10 above)
Comment (tobiasgrosser) [reply SSA/Experimental/Bits/Fast/BitStream.lean]: I implemented this.

File: SSA/Experimental/Bits/Fast/BitStream.lean (hunk 9 above)
Comment (tobiasgrosser) [reply SSA/Experimental/Bits/Fast/BitStream.lean]: This is not needed. @AtticusKuhn showed that mathlib has the same definition.

**Comment by AtticusKuhn** [comment]
> I am done.

Thank you for your guidance. Your feedback is very helpful.

File: SSA/Experimental/Bits/Fast/BitStream.lean [hunk 27]
```diff
@@ -354,12 +354,54 @@ theorem neg_neg : a = - - a := by
...
-theorem ofBitVec_sub : ofBitVec (x - y) ≈ʷ (ofBitVec x) - (ofBitVec y)  := by
-  sorry
+/--
+ Will subAux overflow/carry when computing a - b?
+-/
+def subCarries? (a b : BitStream) (i : Nat) : Bool :=
+  let carry : Bool := match i with
+  | 0 => false
+  | i + 1 => a.subCarries? b i
+  (!a i && b i || !xor (a i) (b i) && carry)
```
Comment (alexkeizer) [inline SSA/Experimental/Bits/Fast/BitStream.lean:333]: I find it hard to follow operator precedence here, with the `xor`. Some more parens would be appreciated!

**Comment by AtticusKuhn** [comment]
If no one objects, then I am going to merge this PR because there has been no activity on it for three days.

File: SSA/Experimental/Bits/Fast/BitStream.lean [hunk 28]
```diff
@@ -320,12 +320,54 @@ theorem neg_neg : a = - - a := by
...
+  induction' i with i ih
+  · simp [subCarries?, negAux, addAux, BitVec.adcb, neg]
+  · by_cases a1 : a (i + 1)
+    <;> by_cases b1 : b (i + 1)
+    <;> cases' @neg_or_add a b i with h h
+    <;> simp [h, subCarries?, ih, negAux, addAux, BitVec.adcb, neg, a1, b1]
+
+theorem subAux_inductive_lemma (i : Nat) :
+    a.subAux b i = ⟨(a.addAux b.neg i).1, subCarries? a b i⟩ := by
+  induction' i with i ih
```
Comment (alexkeizer) [inline SSA/Experimental/Bits/Fast/BitStream.lean:364]: We generally prefer rather mechanical names, based on a theorems type, over names based on intended usage.
This could, e.g., be called `subAux_eq_addAux_neg`

File: SSA/Experimental/Bits/Fast/BitStream.lean [hunk 29]
```diff
@@ -410,6 +452,13 @@ theorem ofBitVec_neg : ofBitVec (- x) ≈ʷ - (ofBitVec x) := by
   _ ≈ʷ ~~~ ofBitVec x   + 1            := add_congr ofBitVec_not_eqTo ofBitVec_one_eqTo_ofNat
   _ ≈ʷ - (ofBitVec x)                  := by rw [neg_eq_not_add]
 
+theorem ofBitVec_sub : ofBitVec (x - y) ≈ʷ (ofBitVec x) - (ofBitVec y)  := by
+  calc
+  _ ≈ʷ ofBitVec (x + -y) := by rw [BitVec.sub_eq_add_neg]
+  _ ≈ʷ ofBitVec x + ofBitVec (-y) := ofBitVec_add
+  _ ≈ʷ ofBitVec x + -(ofBitVec y) := add_congr equal_up_to_refl ofBitVec_neg
+  _ ≈ʷ ofBitVec x - ofBitVec y := by rw [sub_eq_add_neg]
+
```
Comment (alexkeizer) [inline SSA/Experimental/Bits/Fast/BitStream.lean:461]: Gorgeous proof!

File: SSA/Projects/InstCombine/ForLean.lean [hunk 30]
```diff
@@ -633,10 +637,24 @@ theorem not_xor_and_self {a b : Bool} : (!xor a b && b) = (a && b) := by
...
-theorem not_and_self' {a b : Bool} : (a != b && b) = (!a && b) := by
-  by_cases a
-  <;> by_cases b
-  <;> simp [*]
+theorem xor_ne_self {a b : Bool} : xor a ((!a) != b) = !b := by
+  cases a
+  <;> simp
+
+@[simp]
+theorem not_eq_and {a b : Bool} : ((!b) == (a && b)) = (!a && b)  := by
```
Comment (alexkeizer) [inline SSA/Projects/InstCombine/ForLean.lean:651]: You've got an extra whitespace here, just before the `:=`!

**Review by alexkeizer — COMMENTED** [review]
Sorry for the late review! The proof looks nice, thanks for pushing it through
